Separate IsMine from solvability

sipa opened the pull request:
Our current `IsMine` logic does several things with outputs:
* Determine "spendability" (roughly corresponding to "could we sign for this")
* Determine "watching" (is this an output directly or indirectly a watched script)
* Determine invalidity (is this output definitely not legally spendable, detecting accidental uncompressed pubkeys in witnesses)
* Determine "solvability" (would we be able to sign for this ignoring the fact that we may be missing some private keys).

The last item (solvability) is mostly unrelated and only rarely needed (there is just one instance, inside the wallet's coin control logic). This PR changes that instance to use the separate `IsSolvable` function, and stop `IsMine` from distinguishing between solvable and unsolvable.

As an extra, this also simplifies the `IsMine` logic and adds some extra checks (which wouldn't be hit unless someone adds already invalid scripts to their wallet).

sipa committed:
Make coincontrol use IsSolvable to determine solvability

sipa committed:
Make IsMine stop distinguishing solvable/unsolvable

sipa committed:
Simplify IsMine logic

sipa committed:
Add some checks for invalid recursion in IsMine

gmaxwell commented:
Concept ACK

theuni commented on src/script/ismine.cpp:
```diff
@@ -76,6 +76,10 @@ isminetype IsMineInner(const CKeyStore& keystore, const CScript& scriptPubKey, b
     case TX_WITNESS_V0_KEYHASH:
     {
+        if (sigversion == IsMineSigVersion::WITNESS_V0) {
+            // P2WPKH inside P2WSH is invalid.
```
Do these need to flip isInvalid ?

theuni commented:
Concept ACK and light review ACK, but I'm not comfortable enough with this code and its use to ACK confidently.

sipa commented:
@theuni If it helps, all but the last commit should have no effect at all. To see why:
* The first commit rewrites the only code that cares about the distinction between `IsMine` returning `ISMINE_WATCH_UNSOLVABLE` or `ISMINE_WATCH_SOLVABLE` (by calling `IsSolvable` instead, which has essentially the same code as what is inside `IsMine` for distinguishing between the two).
* The second commit makes `IsMine` just stop reporting the difference (and the related constants are renamed).
* The third commit is a simple refactor.

theuni commented on src/wallet/wallet.cpp:
```diff
@@ -2363,10 +2363,10 @@ void CWallet::AvailableCoins(std::vector<COutput> &vCoins, bool fOnlySafe, const
-            bool fSpendableIn = ((mine & ISMINE_SPENDABLE) != ISMINE_NO) || (coinControl && coinControl->fAllowWatchOnly && (mine & ISMINE_WATCH_SOLVABLE) != ISMINE_NO);
-            bool fSolvableIn = (mine & (ISMINE_SPENDABLE | ISMINE_WATCH_SOLVABLE)) != ISMINE_NO;
+            bool solvable = IsSolvable(*this, pcoin->tx->vout[i].scriptPubKey);
+            bool spendable = ((mine & ISMINE_SPENDABLE) != ISMINE_NO) || (coinControl && coinControl->fAllowWatchOnly && solvable);
```
Imo it'd improve readability here to add a check to the second predicate here:
```c++
((mine & ISMINE_WATCH_ONLY) != ISMINE_NONE) && ...
```
While redundant, it's not immediately obvious that invalid watch-only's were filtered out a few lines up. It would also guard against future oopses if a new type is ever added.

sipa replied:
That would be incorrect. When `(coinControl && coinControl->fAllowWatchOnly)` we're expected to set `spendable = true` in `COutPut` for `ISMINE_WATCH_ONLY` outputs.

sipa replied:
Oh, nevermind. I misread the `!=`. Will do.

theuni commented:
@sipa Thanks, that's helpful. I was mostly nervous about bitfield assumptions from the updated isminetype values, but I see now that that's not really a concern.

laanwj commented:
Concept ACK

MeshCollider commented:
Concept ACK

promag commented on src/script/ismine.cpp:
```diff
@@ -42,17 +42,20 @@ bool HaveKeys(const std::vector<valtype>& pubkeys, const CKeyStore& keystore)
+void Update(isminetype& val, isminetype update)
+{
+    if (val == ISMINE_NO) val = update;
+    if (val == ISMINE_WATCH_ONLY && update == ISMINE_SPENDABLE) val = update;
```
`else if`?

jimpo replied:
commit: Simplify IsMine logic

Even better (arguably): `||`.

```c++
if (val == ISMINE_NO ||
    (val == ISMINE_WATCH_ONLY && update == ISMINE_SPENDABLE)) {
  val = update;
}
```

promag reviewed (commented):
IMHO this PR could be split in 2. The change here should be just "Separate IsMine from solvability".

Either way utACK ae572b3

jimpo commented on src/script/ismine.cpp:
```diff
@@ -49,9 +49,9 @@ isminetype IsMineInner(const CKeyStore& keystore, const CScript& scriptPubKey, b
-        if (keystore.HaveWatchOnly(scriptPubKey))
-            return ISMINE_WATCH_UNSOLVABLE;
-        return ISMINE_NO;
+        if (keystore.HaveWatchOnly(scriptPubKey)) {
```
commit: Make IsMine stop distinguishing solvable/unsolvable

Does this even need to be here? Wouldn't it just do the watchonly check at the end of the function anyway?

EDIT: I see it is dropped in the next commit anyway.

jimpo commented on src/script/ismine.cpp:
```diff
@@ -146,23 +144,27 @@ isminetype IsMineInner(const CKeyStore& keystore, const CScript& scriptPubKey, b
 {
-    return IsMineInner(keystore, scriptPubKey, isInvalid, IsMineSigVersion::TOP);
+    isminetype ret = IsMineInner(keystore, scriptPubKey, isInvalid, IsMineSigVersion::TOP);
+    if (isInvalid) {
```
commit: Simplify IsMine logic

It feels like `ISMINE_INVALID` should be an IsMine type. Would require some refactoring of the callers though...

jimpo commented on src/script/ismine.cpp:
```diff
@@ -146,23 +144,27 @@ isminetype IsMineInner(const CKeyStore& keystore, const CScript& scriptPubKey, b
     }
 
-    if (keystore.HaveWatchOnly(scriptPubKey)) {
+    if (ret == ISMINE_NO && keystore.HaveWatchOnly(scriptPubKey)) {
```
commit: Simplify IsMine logic

I get that it's an optimization, but it still strikes me as odd that everywhere else in the function uses `Update` to modify the `ret` value, while this does a check and return.

jimpo reviewed (commented):
Nice set of changes. Only thing is that I don't really like the `Update` pattern -- I find the mutable `ret` value to be harder to reason about in such a complex, recursive function as compared to explicit returns. Especially given that the pattern isn't applied consistently in the final WATCH_ONLY check.

sipa committed:
Make handling of invalid in IsMine more uniform

sipa commented:
Addressed a few comments.

@jimpo: I've added a commit that perhaps makes the data flow clearer, by introducing a separate internal type that does track INVALID as a separate state. For performance reasons, there are still explicit returns in some places. Let me know if you think this helps - I can revert if not.

jimpo commented on src/script/ismine.cpp:
```diff
@@ -28,6 +28,14 @@ enum class IsMineSigVersion
     WITNESS_V0 = 2  //! P2WSH witness script execution
 };
 
+enum class IsMineResult
```
commit: Make handling of invalid in IsMine more uniform

If values were assigned to each (NO = 0, WATCH_ONLY = 1, SPENDABLE = 2, INVALID = 3), `Update` could be implemented as `if (update > val) val = update;`.

sipa replied:
Good idea, done.

jimpo reviewed (commented):
I think the new commit helps in terms of clarity. I still think, though, that isminetype should be merged with IsMineResult (ie. become an enum instead of a bitfield and have an INVALID value). That's a bigger change, so this is fine for now, but I'm curious if you think that's a good direction or whether there's a reason not to do that.

sipa commented:
@jimpo Yes, I think we should do that. I'm planning to first add tester functions (IsMineWatchOnly, IsMineAny, IsMineSpendable, ...) on the bitfield approach now, and then turn it into a real enum.

jimpo commented:
utACK 13142

laanwj commented on src/script/ismine.cpp:
```diff
@@ -42,17 +50,18 @@ bool HaveKeys(const std::vector<valtype>& pubkeys, const CKeyStore& keystore)
+void Update(IsMineResult& val, IsMineResult update)
 {
-    isInvalid = false;
+    if (update > val) val = update;
```
I think the name `Update` is unclear, though I have a hard time thinking of a different term. Effectively it combines two `IsMineResults` and chooses the value with the highest "ownership" (apart from INVALID). Don't know if this would work, or be a good fit:
```c++
Update(ret, IsMineResult::WATCH_ONLY);
```
to
```c++
ret = std::max(ret, IsMineResult::WATCH_ONLY);
```

sipa replied:
Done. The `Update` function was introduced before @jimpo noticed it was really a simple maximum if INVALID is made the highest value.

sipa commented on src/script/ismine.cpp:
```diff
@@ -64,50 +68,58 @@ isminetype IsMineInner(const CKeyStore& keystore, const CScript& scriptPubKey, b
+            return IsMineResult::INVALID;
+        }
+        if (keystore.HaveKey(keyID)) {
+            ret = std::max(ret, IsMineResult::SPENDABLE);
```
I expect the compiler to optimize this out; with the max it's more clear what the intention is (going through various combinations and picking the "most" isminish one, unless it's invalid), and less likely to break with future change.

promag commented on src/script/ismine.cpp:
```diff
@@ -28,6 +28,14 @@ enum class IsMineSigVersion
+    NO = 0,
+    WATCH_ONLY = 1,
+    SPENDABLE = 2,
+    INVALID = 3,
```
nit, add comment to note that the value matters and `INVALID` must be the higher?

sipa replied:
Done, added some comments.

Empact commented on src/wallet/wallet.cpp:
```diff
@@ -2363,10 +2363,10 @@ void CWallet::AvailableCoins(std::vector<COutput> &vCoins, bool fOnlySafe, const
-            bool fSpendableIn = ((mine & ISMINE_SPENDABLE) != ISMINE_NO) || (coinControl && coinControl->fAllowWatchOnly && (mine & ISMINE_WATCH_SOLVABLE) != ISMINE_NO);
-            bool fSolvableIn = (mine & (ISMINE_SPENDABLE | ISMINE_WATCH_SOLVABLE)) != ISMINE_NO;
+            bool solvable = IsSolvable(*this, pcoin->tx->vout[i].scriptPubKey);
+            bool spendable = ((mine & ISMINE_SPENDABLE) != ISMINE_NO) || (((mine & ISMINE_WATCH_ONLY) != ISMINE_NO) && (coinControl && coinControl->fAllowWatchOnly && solvable));
```
nit: this use of ISMINE_NO seems misplaced. Checking for the existence of the flag in mine, the result should not equal 0, which only incidentally is the value of NO.

Empact commented:
utACK c004ffc

laanwj commented:
utACK c004ffc9b42a738043e19e4c812fc7e0566119c5
